server/src/main/java/org/elasticsearch/index/codec/tsdb/TSDBSyntheticIdFieldsProducer.java

```diff
@@ -131,97 +135,371 @@ public boolean hasPayloads() {
-                Uid.decodeId(routingHashBytes.bytes, routingHashBytes.offset, routingHashBytes.length)
+            int tsIdOrdinal = docValues.docTsIdOrdinal(docID);
+            current = new SyntheticTerm(
+                docID,
+                tsIdOrdinal,
+                docValues.lookupTsIdOrd(tsIdOrdinal),
+                docValues.docTimestamp(docID),
+                docValues.docRoutingHash(docID)
             );
-            term = Uid.encodeId(TsidExtractingIdFieldMapper.createId(routingHash, tsId, timestamp));
-            latestTimestamp = timestamp;
-            latestTsId = tsId;
-            return term;
+            return current.term();
         }
 
         @Override
-        public SeekStatus seekCeil(BytesRef id) {
+        public SeekStatus seekCeil(BytesRef id) throws IOException {
+
             assert id != null;
-            if (term != null && term.equals(id)) {
-                return SeekStatus.FOUND;
+            assert Long.BYTES + Integer.BYTES < id.length : id.length;
+            if (id == null || id.length <= Long.BYTES + Integer.BYTES) {
+                return SeekStatus.NOT_FOUND;
             }
-            try {
-                while (next() != null) {
-                    if (term.equals(id)) {
-                        return SeekStatus.FOUND;
+
+            // Extract the _tsid
+            final BytesRef tsId = TsidExtractingIdFieldMapper.extractTimeSeriesIdFromSyntheticId(id);
+            int tsIdOrd = docValues.lookupTsIdTerm(tsId);
+
+            // _tsid not found
+            if (tsIdOrd < 0) {
+                tsIdOrd = -tsIdOrd - 1;
+                // set the terms enum on the first non-matching document
+                if (tsIdOrd < docValues.getTsIdValueCount()) {
+                    int docID = docValues.slowScanToFirstDocWithTsIdOrdinalEqualOrGreaterThan(tsIdOrd);
+                    if (docID != DocIdSetIterator.NO_MORE_DOCS) {
+                        current = new SyntheticTerm(
+                            docID,
+                            tsIdOrd,
+                            docValues.lookupTsIdOrd(tsIdOrd),
+                            docValues.docTimestamp(docID),
+                            docValues.docRoutingHash(docID)
+                        );
+                        return SeekStatus.NOT_FOUND;
                     }
                 }
-            } catch (IOException e) {
-                throw new UncheckedIOException(e);
+                // no docs/terms to iterate on
+                current = NO_MORE_DOCS;
+                return SeekStatus.END;
+            }
+
+            // _tsid found, extract the timestamp
+            final long timestamp = TsidExtractingIdFieldMapper.extractTimestampFromSyntheticId(id);
+
+            // Slow scan to the first document matching the _tsid
+            final int startDocID = docValues.slowScanToFirstDocWithTsIdOrdinalEqualTo(tsIdOrd);
+            assert 0 <= startDocID : startDocID;
```

Comment:
nit: for some reason my brain processes `startDocID >= 0` better than `0 <= startDocID`